# PRACTICA GUIADA 1: Agregando y agrupando datos

* La sumarización es parte fundamental del análisis: computar medidas como ``sum()``, ``mean()``, ``median()``, ``min()`` y ``max()`` son tareas básicas para comenzar a explorar un dataset.

* En esta sección veremos algunas opciones para realizar agregaciones de datos en Pandas.


In [18]:
import numpy as np
import pandas as pd

## Analizando la encuesta en centros de atención turística

El GCBA realiza encuestas a los turistas que se acercan a los centros de atención. Se pregunta el motivo de la consulta, los días que dura el viaje, el país de origen, entre otras cosas.

El dataset es de acceso público en el portal de datos abiertos del GCBA: https://data.buenosaires.gob.ar/dataset/encuesta-en-centros-de-atencion-turistica-cat


In [19]:
turistas = pd.read_csv("resultado-de-encuestas-2016.csv", sep = ";")
turistas.shape

(106942, 53)

In [20]:
turistas.head()

X           Y  ID       FECHA         CAT  \
0  -58,412118  -34,571635   1  14/06/2016  PLANETARIO   
1  -58,412118  -34,571635   2  14/06/2016  PLANETARIO   
2  -58,412118  -34,571635   3  14/06/2016  PLANETARIO   
3  -58,412118  -34,571635   4  14/06/2016  PLANETARIO   
4  -58,412118  -34,571635   5  14/06/2016  PLANETARIO   

                              DIRECCION   BARRIO     COMUNA  PASAJEROS  \
0  AV. SARMIENTO Y AV. FIGUEROA ALCORTA  PALERMO  COMUNA 14          2   
1  AV. SARMIENTO Y AV. FIGUEROA ALCORTA  PALERMO  COMUNA 14          2   
2  AV. SARMIENTO Y AV. FIGUEROA ALCORTA  PALERMO  COMUNA 14          2   
3  AV. SARMIENTO Y AV. FIGUEROA ALCORTA  PALERMO  COMUNA 14          5   
4  AV. SARMIENTO Y AV. FIGUEROA ALCORTA  PALERMO  COMUNA 14          2   

  PERNOCTACIONES       ...        SALUD TANGO  VISITAS_GUIADAS  \
0              7       ...            0     0                0   
1             15       ...            0     0                0   
2              8       ...            0     0                0   
3              7       ...            0     0                0   
4              3       ...            0     0                0   

   ESTUDIANTE_EXTRANJERO  CONSULTA_DE_PASAJERO_EN_TRANSITO  SUBE  APPS  \
0                      0                                 0     0     0   
1                      0                                 0     0     0   
2                      0                                 0     0     0   
3                      0                                 0     0     0   
4                      0                                 0     0     0   

   SURF_URBANO  USO_DE_INSTALACIONES  CANT_DE_MOTIVOS  
0            0                     0                2  
1            0                     0                1  
2            0                     0                3  
3            0                     0                7  
4            0                     0                1  

[5 rows x 53 columns]

Veamos todas las columnas

In [21]:
turistas.columns

Index(['X', 'Y', 'ID', 'FECHA', 'CAT', 'DIRECCION', 'BARRIO', 'COMUNA',
       'PASAJEROS', 'PERNOCTACIONES', 'ARGENTINA', 'PAIS_EXT', 'ACCESIBILIDAD',
       'AGENCIAS_COMPAÑIAS_AEREAS', 'AGENDA_CULTURAL', 'ALOJAMIENTO',
       'ALREDEDORES_BA', 'ARTE_Y_ARQUITECTURA', 'ATRACTIVOS_TURISTICOS',
       'CAMBIO_BANCO_CAJEROS', 'BARRIOS', 'BICICLETAS', 'BUS_TURISTICO',
       'COMPRAS', 'DEPORTES', 'EVENTOS_ESPECTACULOS', 'FERIAS_MERCADOS',
       'GASTRONOMIA', 'GUIAS_TEMATICAS', 'INFO_ARGENTINA', 'INFO_URUGUAY',
       'LGBTIQ', 'LITERARIO', 'MAPAS_CIUDAD', 'MUSEOS', 'NIÑOS',
       'VIDA_NOCTURNA', 'ORGANISMOS_OFICIALES', 'ORIENTACION_TRANSPORTE',
       'OTROS', 'PARQUES', 'QUE_HACER_EN_BA', 'RELIGIOSO', 'SALUD', 'TANGO',
       'VISITAS_GUIADAS', 'ESTUDIANTE_EXTRANJERO',
       'CONSULTA_DE_PASAJERO_EN_TRANSITO', 'SUBE', 'APPS', 'SURF_URBANO',
       'USO_DE_INSTALACIONES', 'CANT_DE_MOTIVOS'],
      dtype='object')

## Agregaciones simples en Pandas

* Al igual que en un array de una dimensión para una ``Series`` la agregación devuelve un valor único. Veamos cuántos días en total y en promedio dura el viaje de las personas que se acercaron a estos centros de atención.

In [22]:
turistas["PERNOCTACIONES"]

0             7
1            15
2             8
3             7
4             3
5             8
6             4
7         NS/NC
8            11
9             4
10           21
11           12
12            5
13            2
14            5
15            2
16           10
17            4
18            6
19            7
20        NS/NC
21        NS/NC
22           12
23            3
24            3
25        NS/NC
26           10
27        NS/NC
28           15
29            5
          ...  
106912        1
106913        1
106914        2
106915        1
106916        1
106917        1
106918        1
106919        1
106920        1
106921        3
106922        2
106923        1
106924    NS/NC
106925        1
106926       14
106927    NS/NC
106928        5
106929        7
106930        2
106931    NS/NC
106932        1
106933        1
106934        1
106935    NS/NC
106936    NS/NC
106937       15
106938        7
106939       10
106940    NS/NC
106941       17
Name: PERNOCTACIONES, Le

Veamos que la columna incluye valores no numéricos, saquémoslos y luego hagamos los cálculos.

In [23]:
# el método str.extract permite extraer un grupo pasando una REGEX. Para probar REGEX se puede entrar en https://regex101.com/ o https://regexr.com/ 
# expand lo convierte en una serie de pandas cuando es TRUE

turistas['PERNOCTACIONES'] = turistas['PERNOCTACIONES'].str.extract('(\d+)', expand=False).astype('float')

In [24]:
turistas["PERNOCTACIONES"].sum()

491096.0

In [25]:
turistas["PERNOCTACIONES"].mean()

6.940305257207462

* Para un ``DataFrame``, los `aggregate` retornan por default resultados al interior de cada columna:

* Especificando el argumento ``axis`` es posible agregar los resultados para cada fila.

**Nota:** Al igual que en el caso de NumPy, en `axis` se define la dimensión sobre la que se colapsan los datos. Si quiero resultados por fila, entonces, `axis='columns'` y en otro caso `axis='rows'`

En el archivo ya viene calculada la cantidad de motivos por los que se acercan los turistas. Calculemos nosotros ese dato y comparémoslo con el original, ya que es común que hayan errores en la información.

In [26]:
# Los motivos van desde la columna ACCESIBILIDAD hasta USO_DE_INSTALACIONES
# Calculo la suma por fila
n_motivos_calculado = turistas.loc[:,"ACCESIBILIDAD":"USO_DE_INSTALACIONES"].sum(axis = 'columns')
# Me quedo con el valor original
n_motivos_original = turistas['CANT_DE_MOTIVOS']

# Vemos que hay errores en el campo original:
print('¿Cantidad total calculada y original son iguales? {}'.format(n_motivos_calculado.equals(n_motivos_original)))

¿Cantidad total calculada y original son iguales? False


In [27]:
# ¿Cuántos están mal? 52
(n_motivos_calculado != n_motivos_original).sum()

52

* Las ``Series`` y ``DataFrame`` de Pandas incluyen muchas funciones de agregación.

* Hay, además, un método ``describe()`` que computa algunas medidas agregadas por defecto para cada columna y devuelve un `DataFrame` como resultado.


In [28]:
turistas[['PASAJEROS', 'PERNOCTACIONES']].describe()

PASAJEROS  PERNOCTACIONES
count  106942.000000    70760.000000
mean        1.961764        6.940305
std         1.327208       20.220092
min         1.000000        0.000000
25%         1.000000        3.000000
50%         2.000000        4.000000
75%         2.000000        6.000000
max       150.000000      365.000000

Podemos observar que hay valores raros, ya que en pasajeros vemos 150... ¿es un error o nos falta conocer "el negocio"? Tal vez alguien consulto por un grupo de viajeros...

In [29]:
turistas[['PASAJEROS', 'PERNOCTACIONES']].dropna().describe()

PASAJEROS  PERNOCTACIONES
count  70760.000000    70760.000000
mean       2.061885        6.940305
std        1.411212       20.220092
min        1.000000        0.000000
25%        1.000000        3.000000
50%        2.000000        4.000000
75%        2.000000        6.000000
max      150.000000      365.000000

In [30]:
turistas['PERNOCTACIONES'].isnull().sum()

36182

* ¿Qué pueden decir de estos datos?

In [31]:
# Alternativas para el método .desribe()

turistas.describe(percentiles=[0.1,0.25,0.35,0.5,0.8], include='float')

PERNOCTACIONES
count    70760.000000
mean         6.940305
std         20.220092
min          0.000000
10%          2.000000
25%          3.000000
35%          3.000000
50%          4.000000
80%          7.000000
max        365.000000

Resumen de algunas funciones de agregación en Pandas:

| Aggregation              | Description                     |
|--------------------------|---------------------------------|
| ``count()``              | Total number of items           |
| ``mean()``, ``median()`` | Mean and median                 |
| ``min()``, ``max()``     | Minimum and maximum             |
| ``std()``, ``var()``     | Standard deviation and variance |
| ``mad()``                | Mean absolute deviation         |
| ``prod()``               | Product of all items            |
| ``sum()``                | Sum of all items                |

Todos son métodos de los objetos ``DataFrame`` y ``Series``.

* Muchas veces esto no es suficiente y es necesario usar otras operaciones de sumarización.
* La operación ``groupby`` permite computar medidas agregadas de forma eficente en subsets de los datos.

## GroupBy: Split, Apply, Combine

* Muchas veces es importante poder realizar operaciones de agregación de forma condicional a algún subconjunto de datos (por ejemplo, para los casos que cumplen alguna condición). Esto es implementado por el operador `groupby`
* El nombre `groupby` proviene del lenguaje SQL.
* Es útil pensarlo en los términos de Hadley Wickham: *split, apply, combine* (dividir-aplicar-combinar)

### Split, apply, combine

![title](img/split-apply-combine.png)

- El paso *split*  implica dividir y reagrupar un ``DataFrame`` en base a una determinada key.
- El paso *apply* supone computar alguna función (generalmente, alguna agregación, transformación o filtro) sobre los grupos constituidos en el paso anterior.
- El paso *combine* hace un "merge" de los resultados de dichas operaciones en un nuevo array.

* Si bien cada uno de los pasos pueden hacerse "manualmente", el ``GroupBy`` generalmente puede hacerlo en un solo paso.
* La ventaja del ``GroupBy`` es que permite abstraer los tres pasos anteriores: el usuario no necesita pensar en "cómo" hacer el cómputo, sino más bien pensar la operación como un todo.

* La operación split-apply-combine más básica con el método ``groupby()`` es pasar el nombre de una determinada clave de columna:

### El objeto GroupBy

* El objeto ``GroupBy`` es un abstracción muy flexible.
* Puede ser tratado como una colección de `DataFrame` y hace algunas cosas complicadas por detrás...
* Quizá las operaciones más importantes de un objeto ``GroupBy`` sean las *aggregate*, *filter*, *transform*, y *apply*

In [32]:
turistas.groupby('PAIS_EXT')

* Notar que no se retorna un ``DataFrame`` sino un objeto ``DataFrameGroupBy``.
* Se puede pensar en el `DataFrameGroupBy` como una vista especial de un `DataFrame` que construye los grupos pero que no realiza ningún cómputo hasta que la etapa de agregación es aplicada.
* Esta **"lazy evaluation"** hace que las operaciones de agregación puedan ser computadas de forma muy eficiente.
* Para producir un resultado podemos aplicar una función de agregación a este ``DataFrameGroupBy`` que a realizar la operación (apply) y avanzar en el paso de combine.

In [34]:
## RAMA se ponene 2 corchetes para cuando queremos usar mas de una serie y los 2 corchetes lo transforman en dataframe

turistas[['PASAJEROS', 'PERNOCTACIONES','PAIS_EXT']].groupby('PAIS_EXT').sum().sort_values("PASAJEROS", 
                                                                                           ascending = False).head()

PASAJEROS  PERNOCTACIONES
PAIS_EXT                                 
                    70292         78226.0
BRASIL              41810         78091.0
CHILE               12262         27264.0
ESTADOS UNIDOS      11032         29960.0
ESPAÑA               8728         31087.0

* Se podría realizar virtualmente cualquier operación común de agregación de Pandas o NumPy y casi cualquier operación válida para un `DataFrame` como veremos a continuación.

#### Indexing de columnas

* El objeto ``GroupBy`` soporta el indexado de columnas de la misma forma que un ``DataFrame``, y devuelve un objeto ``GroupBy`` modificado:

In [35]:
turistas.groupby('PAIS_EXT')['ALREDEDORES_BA']

* Seleccionamos una ``Series`` particular del  ``DataFrame`` original referenciándolo con su nombre de columna.
* Al igual que antes, no se realiza ningún cómputo hasta que llamamos alguna función de agregación sobre el objeto.

In [38]:
turistas.groupby('PAIS_EXT')['ALREDEDORES_BA'].sum().sort_values(ascending = False).head()

PAIS_EXT
           546
BRASIL     382
CHILE      163
ESPAÑA     126
URUGUAY    108
Name: ALREDEDORES_BA, dtype: int64

* ¿Qué se observa en estos datos?

#### Iteración sobre grupos

* El objeto ``GroupBy`` soporta la iteración directa sobre grupos, devolviendo cada grupo como una ``Series`` o un ``DataFrame``.


In [43]:
for (pais, group) in turistas.groupby('PAIS_EXT'):
    print("{0:20} shape={1} ".format(pais, group.shape))

                     shape=(39685, 53) 
ALBANIA              shape=(1, 53) 
ALEMANIA             shape=(2501, 53) 
ANDORRA              shape=(1, 53) 
ANGOLA               shape=(4, 53) 
ARABIA SAUDITA       shape=(2, 53) 
ARGELIA              shape=(1, 53) 
ARMENIA              shape=(12, 53) 
ARUBA                shape=(2, 53) 
AUSTRALIA            shape=(1474, 53) 
AUSTRIA              shape=(181, 53) 
AZERBAIYAN           shape=(1, 53) 
BAHAMAS              shape=(4, 53) 
BANGLADESH           shape=(1, 53) 
BELGICA              shape=(193, 53) 
BERMUDAS             shape=(1, 53) 
BIELORRUSIA          shape=(3, 53) 
BOLIVIA              shape=(525, 53) 
BOSNIA Y HERZEGOVINA shape=(1, 53) 
BRASIL               shape=(18613, 53) 
BRUNEI               shape=(1, 53) 
BULGARIA             shape=(16, 53) 
CABO VERDE           shape=(1, 53) 
CANADA               shape=(1165, 53) 
CATAR                shape=(3, 53) 
CHILE                shape=(5630, 53) 
CHINA                shape=(636, 53)

* Si bien podemos hacer este tipo de operaciones de forma manual, veremos enseguida la potencialidad que tienen las funcionalidades ``apply``.

#### Dispatch methods

* Cualquier método no implementado de forma explícita por el objeto ``GroupBy`` será ejecutado en cada grupo. Estos métodos se heredan de la clase DataFrame. 
* Por ejemplo, se puede usar el método ``describe()`` de ``DataFrame``s para realizar muchas operaciones de agregación al interior de cada grupo.

In [44]:
turistas.groupby('PAIS_EXT')[['PASAJEROS', 'PERNOCTACIONES']].describe()

PASAJEROS                                            \
                         count      mean       std  min   25%  50%   75%   
PAIS_EXT                                                                   
                       39685.0  1.771249  1.316617  1.0  1.00  2.0  2.00   
ALBANIA                    1.0  2.000000       NaN  2.0  2.00  2.0  2.00   
ALEMANIA                2501.0  1.834866  0.950537  1.0  1.00  2.0  2.00   
ANDORRA                    1.0  1.000000       NaN  1.0  1.00  1.0  1.00   
ANGOLA                     4.0  1.750000  0.957427  1.0  1.00  1.5  2.25   
ARABIA SAUDITA             2.0  1.500000  0.707107  1.0  1.25  1.5  1.75   
ARGELIA                    1.0  4.000000       NaN  4.0  4.00  4.0  4.00   
ARMENIA                   12.0  2.166667  0.937437  1.0  2.00  2.0  2.00   
ARUBA                      2.0  1.500000  0.707107  1.0  1.25  1.5  1.75   
AUSTRALIA               1474.0  2.023066  0.946360  1.0  1.00  2.0  2.00   
AUSTRIA                  181.0  1.845304  1.037273  1.0  1.00  2.0  2.00   
AZERBAIYAN                 1.0  2.000000       NaN  2.0  2.00  2.0  2.00   
BAHAMAS                    4.0  2.500000  1.732051  1.0  1.75  2.0  2.75   
BANGLADESH                 1.0  2.000000       NaN  2.0  2.00  2.0  2.00   
BELGICA                  193.0  1.844560  0.807958  1.0  1.00  2.0  2.00   
BERMUDAS                   1.0  1.000000       NaN  1.0  1.00  1.0  1.00   
BIELORRUSIA                3.0  1.666667  0.577350  1.0  1.50  2.0  2.00   
BOLIVIA                  525.0  2.001905  1.124839  1.0  1.00  2.0  2.00   
BOSNIA Y HERZEGOVINA       1.0  1.000000       NaN  1.0  1.00  1.0  1.00   
BRASIL                 18613.0  2.246279  1.238076  1.0  2.00  2.0  2.00   
BRUNEI                     1.0  2.000000       NaN  2.0  2.00  2.0  2.00   
BULGARIA                  16.0  1.625000  0.718795  1.0  1.00  1.5  2.00   
CABO VERDE                 1.0  2.000000       NaN  2.0  2.00  2.0  2.00   
CANADA                  1165.0  2.043777  1.807955  1.0  1.00  2.0  2.00   
CATAR                      3.0  3.333333  2.309401  2.0  2.00  2.0  4.00   
CHILE                   5630.0  2.177975  2.355871  1.0  2.00  2.0  2.00   
CHINA                    636.0  1.888365  1.169880  1.0  1.00  2.0  2.00   
COLOMBIA                3545.0  2.045980  1.148596  1.0  1.00  2.0  2.00   
COREA DEL NORTE           10.0  1.800000  0.788811  1.0  1.00  2.0  2.00   
COREA DEL SUR            376.0  1.816489  1.035809  1.0  1.00  2.0  2.00   
...                        ...       ...       ...  ...   ...  ...   ...   
REINO UNIDO             2132.0  1.963884  1.084903  1.0  1.00  2.0  2.00   
REPUBLICA CHECA           43.0  1.790698  0.989424  1.0  1.00  2.0  2.00   
REPUBLICA DEL CONGO        1.0  1.000000       NaN  1.0  1.00  1.0  1.00   
REPUBLICA DOMINICANA      38.0  2.131579  1.166447  1.0  1.00  2.0  2.00   
RUMANIA                   35.0  1.657143  0.764771  1.0  1.00  2.0  2.00   
RUSIA                    230.0  1.930435  1.534020  1.0  1.00  2.0  2.00   
SAN MARINO                 2.0  2.000000  0.000000  2.0  2.00  2.0  2.00   
SENEGAL                    2.0  1.500000  0.707107  1.0  1.25  1.5  1.75   
SERBIA                    15.0  1.400000  0.507093  1.0  1.00  1.0  2.00   
SINGAPUR                  35.0  1.571429  0.739066  1.0  1.00  1.0  2.00   
SIRIA                      5.0  1.800000  0.447214  1.0  2.00  2.0  2.00   
SOMALIA                    1.0  2.000000       NaN  2.0  2.00  2.0  2.00   
SRI LANKA                  1.0  1.000000       NaN  1.0  1.00  1.0  1.00   
SUDAFRICA                 73.0  2.301370  1.515521  1.0  2.00  2.0  2.00   
SUECIA                   156.0  1.884615  0.976905  1.0  1.00  2.0  2.00   
SUIZA                    597.0  1.879397  0.968737  1.0  1.00  2.0  2.00   
SURINAM                    2.0  2.000000  1.414214  1.0  1.50  2.0  2.50   
TAILANDIA                 19.0  2.368421  1.011628  1.0  2.00  2.0  3.00   
TAIWAN                    29.0  1.793103  1.235756  1.0  1.00  1.0  2

* ¿Qué puede decirse de estos datos? 
* Este es solo un ejemplo de la utilidad de estos métodos. Notar que son aplicados a cada uno de los grupos individuales y que los resultados son combinados en un objeto `GroupBy` y retornados.

### Aggregate, filter, transform, apply

* Hay muchas más opciones de operaciones disponibles.
* Los objetos ``GroupBy`` tienen algunos métodos muy útiles: ``aggregate()``, ``filter()``, ``transform()`` y ``apply()`` que implementan muchas operaciones en la etapa previa al "combine"

* Ilustremos estas operaciones con el siguiente ``DataFrame``:

#### Aggregation

* El método ``aggregate()`` permite una gran flexibilidad:
* Puede tomar un string, una función o una lista y computar todos los agregados en un solo paso.

In [45]:
turistas.groupby('ARGENTINA')["PASAJEROS"].aggregate([min, np.mean, np.median, max, sum, ])

min      mean  median  max     sum
ARGENTINA                                                 
                          1  2.073941       2  150  140130
BUENOS AIRES              1  2.021074       2   50    9111
CATAMARCA                 1  1.793750       2    9     287
CHACO                     1  1.876751       2    7     670
CHUBUT                    1  1.899371       2    5     604
CIUDAD DE BUENOS AIRES    1  1.510368       1   40   31174
CORDOBA                   1  2.119264       2   25    9098
CORRIENTES                1  1.920502       2   12     918
ENTRE RIOS                1  1.927210       2   19    1112
FORMOSA                   1  1.892617       2   10     282
JUJUY                     1  2.036585       2   22     501
LA PAMPA                  1  1.883495       2    5     388
LA RIOJA                  1  1.764706       2    4     180
MENDOZA                   1  2.123726       2   60    2918
MISIONES                  1  1.897690       2    7     575
NEUQUEN                   1  2.350515       2   60    1140
NO INFORMO PROVINCIA      1  1.166667       1    2       7
RIO NEGRO                 1  2.070769       2   50     673
SALTA                     1  1.931034       2    6     840
SAN JUAN                  1  2.065421       2   41     663
SAN LUIS                  1  1.979424       2   10     481
SANTA CRUZ                1  1.953704       2    6     422
SANTA FE                  1  2.104964       2   45    5555
SANTIAGO DEL ESTERO       1  2.028571       2    5     284
TIERRA DEL FUEGO          1  2.005714       2    5     351
TUCUMAN                   1  2.107511       2   20    1431

* Otra forma útil es pasar un diccionario que mapea nombres de columnas con operaciones. De esta forma se puede espeficiar una operación distinta a cada columna.

In [46]:
turistas.groupby('ARGENTINA')['PASAJEROS'].aggregate([min, 
                 np.median, max, sum]).rename(columns={'min': 'grupo_mas_chico',
                'median': 'grupo_mediano','max':'grupo_mas_chico','sum':'total'}).head()

grupo_mas_chico  grupo_mediano  grupo_mas_chico   total
ARGENTINA                                                            
                            1              2              150  140130
BUENOS AIRES                1              2               50    9111
CATAMARCA                   1              2                9     287
CHACO                       1              2                7     670
CHUBUT                      1              2                5     604

#### Filtering

* Una operación filtering permite "descartar" datos basado en propiedades del grupo.
* Por ejemplo, podríamos querer mantener todos los grupos en los que la desviación estándar sea mayor que algún valor de corte:

In [47]:
def filter_func(x):
    return x["PERNOCTACIONES"].std() > 20

In [48]:
filter_func(turistas)

True

In [49]:
turistas.groupby('ARGENTINA').filter(filter_func)["ARGENTINA"].unique()

array([' ', 'CIUDAD DE BUENOS AIRES', 'SALTA', 'CHACO', 'SAN JUAN',
       'TUCUMAN', 'JUJUY', 'TIERRA DEL FUEGO', 'CORRIENTES'], dtype=object)

* La función de filtro retorna un booleano especificando si el grupo pasa o no el filtro. 

#### Transformation

* Mientras que aggregation devuelve una versión reducida de los datos, transformation retorna alguna versión transformada de los datos para, luego, hacer el combine.
* El output de una transformation es del mismo `shape` que el input.

In [50]:
def center_mean(x):
    return(x - x.mean())

print(turistas[["ARGENTINA","PASAJEROS"]].groupby('ARGENTINA').transform(center_mean))

        PASAJEROS
0       -0.073941
1       -0.073941
2       -0.073941
3        2.876274
4       -0.119264
5       -0.073941
6       -0.073941
7        3.489632
8       -0.073941
9       -0.073941
10      -1.073941
11       1.926059
12      -0.073941
13      -0.021074
14       2.880736
15      -0.119264
16       1.926059
17       0.876274
18      -0.073941
19       0.926059
20       0.489632
21      -0.510368
22      -0.073941
23      -0.073941
24       0.068966
25       2.489632
26      -0.073941
27      -0.510368
28      -0.350515
29      -0.073941
...           ...
106912  -1.021074
106913  -0.104964
106914  -0.021074
106915  -0.119264
106916  -0.119264
106917  -0.021074
106918  -0.104964
106919  -0.119264
106920  -1.119264
106921  -0.021074
106922  -0.119264
106923  -0.119264
106924  -0.510368
106925  -1.104964
106926  -1.104964
106927  -1.073941
106928  -1.073941
106929  -1.073941
106930  -1.073941
106931  -1.073941
106932   1.926059
106933  -1.119264
106934  -0.104964
106935  -0

* ¿Qué hace el siguiente bloque de código?

#### El método `apply()`

* El método `apply()` permite aplicar alguna función dada a los resultados del group.
* La función debería tomar como input ``DataFrame`` y devolver un objeto de Pandas (``DataFrame``, ``Series``) o un escalar. 
* La operación combine se adaptará al tipo de salida devuelta.

In [51]:
# El resultado es una columna PASAJEROS normalizada por la participación 
#del grupo en el total de pasajeros de la provincia:

def norm_by_data2(x):
    # x is a DataFrame of group values
    x['PASAJEROS'] /= x['PERNOCTACIONES'].sum()
    return x

turistas[["ARGENTINA","PASAJEROS","PERNOCTACIONES"]].groupby('ARGENTINA').apply(norm_by_data2).sample(15)


ARGENTINA  PASAJEROS  PERNOCTACIONES
44371                            0.000005             5.0
100236  CIUDAD DE BUENOS AIRES   0.005479             NaN
94082   CIUDAD DE BUENOS AIRES   0.002740             NaN
91208                            0.000005             NaN
51259                            0.000005             3.0
66289   CIUDAD DE BUENOS AIRES   0.002740             NaN
77232                  NEUQUEN   0.000760             3.0
62023                            0.000005             8.0
100842                           0.000005             NaN
24731                            0.000007             5.0
55287                            0.000005             4.0
83303   CIUDAD DE BUENOS AIRES   0.005479             NaN
57430                            0.000005             3.0
73784                            0.000005             NaN
103665                           0.000010             4.0

* ¿Qué operación produce el bloque de código anterior?

### Especificando la clave del "split"

* En el ejemplo anterior se hacía el split del ``DataFrame`` sobre una sola columna.
* Hay otras opciones...

#### Una lista, array, series, o index que contiene las claves del grouping 

In [52]:
L = ["CAT","ARGENTINA"]
turistas.groupby(L).sum()

ID  PASAJEROS  PERNOCTACIONES  \
CAT        ARGENTINA                                                      
AEROPARQUE                          80696414       2509         12120.0   
           BUENOS AIRES              7941050        224           342.0   
           CATAMARCA                 1244500         40           102.0   
           CHACO                      983903         34            94.0   
           CHUBUT                    1993985         49           216.0   
           CIUDAD DE BUENOS AIRES   85900646       1958             0.0   
           CORDOBA                  10078375        329           589.0   
           CORRIENTES                1354056         35           106.0   
           ENTRE RIOS                 601956         14            60.0   
           FORMOSA                    608369         12            42.0   
           JUJUY                     1494666         42           451.0   
           LA PAMPA                   459164         13            27.0   
           LA RIOJA                   470381          9            28.0   
           MENDOZA                   5094207        153           562.0   
           MISIONES                  1694579         38           118.0   
           NEUQUEN                   3307433        107           297.0   
           RIO NEGRO                 1981710         68           296.0   
           SALTA                     3940129        124           741.0   
           SAN JUAN                  1382512         45            87.0   
           SAN LUIS                   526583         13            31.0   
           SANTA CRUZ                1414308         41           245.0   
           SANTA FE                  1541733         41            89.0   
           SANTIAGO DEL ESTERO        300958         10            12.0   
           TIERRA DEL FUEGO          1157823         27           107.0   
           TUCUMAN                   2989178        114           261.0   
CAMINITO                           327420176      11039         22340.0   
           BUENOS AIRES             18208809        926          1007.0   
           CATAMARCA                   81517          1             0.0   
           CHACO                      764989         42            83.0   
           CHUBUT                     721815         29            63.0   
...                                      ...        ...             ...   
RECOLETA   SANTA CRUZ                 547779         32            81.0   
           SANTA FE                  9532511        637          1038.0   
           SANTIAGO DEL ESTERO        321942         19            53.0   
           TIERRA DEL FUEGO           790664         44           102.0   
           TUCUMAN                   2318074        125           305.0   
RETIRO                             288043065       8751         36798.0   
           BUENOS AIRES             48819924       1429          2243.0   
           CATAMARCA                 3281308         85           258.0   
           CHACO                     8212223        230          1728.0   
           CHUBUT                    2395694         75           251.0   
           CIUDAD DE BUENOS AIRES   62675674       1505           365.0   
           CORDOBA                  53653823       1603          2723.0   
           CORRIENTES               10490033        283          1193.0   
           ENTRE RIOS               12514477        358           599.0   
           FORMOSA                   3431179        103           459.0   
           JUJUY                     2945946         98           518.0   
           LA PAMPA                  4674995        115           220.0   
           LA RIOJA                  2222437         65           170.0   
           MENDOZA                  16467762        549          1113.0   
           MISIONES                  6993580        200           692.0   
           NEUQUEN                   3309306        103           2

In [53]:
turistas.groupby("ARGENTINA").sum().head()


ID  PASAJEROS  PERNOCTACIONES  ACCESIBILIDAD  \
ARGENTINA                                                            
              3611155388     140130        413915.0             68   
BUENOS AIRES   241442709       9111         10968.0              1   
CATAMARCA       10091162        287           671.0              0   
CHACO           18178772        670          2791.0              0   
CHUBUT          16568384        604          1806.0              0   

              AGENCIAS_COMPAÑIAS_AEREAS  AGENDA_CULTURAL  ALOJAMIENTO  \
ARGENTINA                                                               
                                    655             1440         1119   
BUENOS AIRES                         28              178           63   
CATAMARCA                             2                5            5   
CHACO                                 1               21           12   
CHUBUT                                2               17            6   

              ALREDEDORES_BA  ARTE_Y_ARQUITECTURA  ATRACTIVOS_TURISTICOS  \
ARGENTINA                                                                  
                        1642                 1128                   7340   
BUENOS AIRES              66                   48                    328   
CATAMARCA                  2                    0                      9   
CHACO                      9                    3                     19   
CHUBUT                    13                    6                     32   

                   ...         SALUD  TANGO  VISITAS_GUIADAS  \
ARGENTINA          ...                                         
                   ...            85   1986             1188   
BUENOS AIRES       ...            20    112              155   
CATAMARCA          ...             1      8                4   
CHACO              ...             4     19                3   
CHUBUT             ...             0     12                3   

              ESTUDIANTE_EXTRANJERO  CONSULTA_DE_PASAJERO_EN_TRANSITO  SUBE  \
ARGENTINA                                                                     
                                 25                                95  2212   
BUENOS AIRES                      0                                11   186   
CATAMARCA                         0                                 1    10   
CHACO                             0                                 2    25   
CHUBUT                            0                                 1     9   

              APPS  SURF_URBANO  USO_DE_INSTALACIONES  CANT_DE_MOTIVOS  
ARGENTINA                                                               
                34           80                   366           134473  
BUENOS AIRES     0            3                    24             7624  
CATAMARCA        0            0                     1              295  
CHACO            0            1                     3              690  
CHUBUT           0            2                     0              608  

[5 rows x 44 columns]

In [54]:
# Es totalmente equivaliente a
turistas.groupby(turistas["ARGENTINA"]).sum().head()


ID  PASAJEROS  PERNOCTACIONES  ACCESIBILIDAD  \
ARGENTINA                                                            
              3611155388     140130        413915.0             68   
BUENOS AIRES   241442709       9111         10968.0              1   
CATAMARCA       10091162        287           671.0              0   
CHACO           18178772        670          2791.0              0   
CHUBUT          16568384        604          1806.0              0   

              AGENCIAS_COMPAÑIAS_AEREAS  AGENDA_CULTURAL  ALOJAMIENTO  \
ARGENTINA                                                               
                                    655             1440         1119   
BUENOS AIRES                         28              178           63   
CATAMARCA                             2                5            5   
CHACO                                 1               21           12   
CHUBUT                                2               17            6   

              ALREDEDORES_BA  ARTE_Y_ARQUITECTURA  ATRACTIVOS_TURISTICOS  \
ARGENTINA                                                                  
                        1642                 1128                   7340   
BUENOS AIRES              66                   48                    328   
CATAMARCA                  2                    0                      9   
CHACO                      9                    3                     19   
CHUBUT                    13                    6                     32   

                   ...         SALUD  TANGO  VISITAS_GUIADAS  \
ARGENTINA          ...                                         
                   ...            85   1986             1188   
BUENOS AIRES       ...            20    112              155   
CATAMARCA          ...             1      8                4   
CHACO              ...             4     19                3   
CHUBUT             ...             0     12                3   

              ESTUDIANTE_EXTRANJERO  CONSULTA_DE_PASAJERO_EN_TRANSITO  SUBE  \
ARGENTINA                                                                     
                                 25                                95  2212   
BUENOS AIRES                      0                                11   186   
CATAMARCA                         0                                 1    10   
CHACO                             0                                 2    25   
CHUBUT                            0                                 1     9   

              APPS  SURF_URBANO  USO_DE_INSTALACIONES  CANT_DE_MOTIVOS  
ARGENTINA                                                               
                34           80                   366           134473  
BUENOS AIRES     0            3                    24             7624  
CATAMARCA        0            0                     1              295  
CHACO            0            1                     3              690  
CHUBUT           0            2                     0              608  

[5 rows x 44 columns]

#### Un dict o serie mapeando index a grupos

In [55]:
turistas2 = turistas.set_index('ARGENTINA')
mapping = {'CHACO': 'C', 'CHUBUT': 'D', 'CATAMARCA': 'E'}
turistas2.groupby(mapping).sum()

ID  PASAJEROS  PERNOCTACIONES  ACCESIBILIDAD  \
C  18178772        670          2791.0              0   
D  16568384        604          1806.0              0   
E  10091162        287           671.0              0   

   AGENCIAS_COMPAÑIAS_AEREAS  AGENDA_CULTURAL  ALOJAMIENTO  ALREDEDORES_BA  \
C                          1               21           12               9   
D                          2               17            6              13   
E                          2                5            5               2   

   ARTE_Y_ARQUITECTURA  ATRACTIVOS_TURISTICOS       ...         SALUD  TANGO  \
C                    3                     19       ...             4     19   
D                    6                     32       ...             0     12   
E                    0                      9       ...             1      8   

   VISITAS_GUIADAS  ESTUDIANTE_EXTRANJERO  CONSULTA_DE_PASAJERO_EN_TRANSITO  \
C                3                      0                                 2   
D                3                      0                                 1   
E                4                      0                                 1   

   SUBE  APPS  SURF_URBANO  USO_DE_INSTALACIONES  CANT_DE_MOTIVOS  
C    25     0            1                     3              690  
D     9     0            2                     0              608  
E    10     0            0                     1              295  

[3 rows x 44 columns]

### Ejemplo de aplicación

* Imaginemos que nos consultan cuáles son los lugares de CABA más visitados por los turistas argentinos, según su procedencia. Para ello suponemos que el CAT al que van los turistas indica que fueron a visitar ese lugar. 
* Entonces primero filtramos los turistas extranjeros, agrupamos por provincia, llenamos los Na con 0 y dividimos todo por la cantidad total de turistas argentinos. De esta manera, lo representamos como proporción del total

In [56]:
(turistas.loc[turistas.ARGENTINA != " ",].groupby(['ARGENTINA', 
                      'CAT'])['PASAJEROS'].sum().unstack().fillna(0)/sum(turistas.loc[turistas.ARGENTINA != " ",
                                                                                            "PASAJEROS"])*100)

CAT                     AEROPARQUE  CAMINITO    FLORIDA       HUB  PLANETARIO  \
ARGENTINA                                                                       
BUENOS AIRES              0.321539  1.329218   3.821144  0.173688    1.165578   
CATAMARCA                 0.057418  0.001435   0.126319  0.000000    0.015790   
CHACO                     0.048805  0.060289   0.223929  0.000000    0.031580   
CHUBUT                    0.070337  0.041628   0.268427  0.002871    0.104787   
CIUDAD DE BUENOS AIRES    2.810594  2.812029  14.421876  0.228235    4.576186   
CORDOBA                   0.472260  0.849781   3.747937  0.044499    0.782315   
CORRIENTES                0.050240  0.053111   0.401923  0.005742    0.040192   
ENTRE RIOS                0.020096  0.083256   0.380392  0.002871    0.064595   
FORMOSA                   0.017225  0.010048   0.080385  0.000000    0.030144   
JUJUY                     0.060289  0.063159   0.225364  0.000000    0.022967   
LA PAMPA                  0.018661  0.010048   0.222493  0.000000    0.014354   
LA RIOJA                  0.012919  0.000000   0.071772  0.000000    0.005742   
MENDOZA                   0.219622  0.302878   1.192852  0.015790    0.192349   
MISIONES                  0.054547  0.035886   0.192349  0.000000    0.050240   
NEUQUEN                   0.153592  0.170817   0.348812  0.011484    0.078949   
NO INFORMO PROVINCIA      0.000000  0.000000   0.002871  0.000000    0.000000   
RIO NEGRO                 0.097610  0.063159   0.206704  0.002871    0.081820   
SALTA                     0.177995  0.063159   0.360296  0.007177    0.048805   
SAN JUAN                  0.064595  0.101916   0.218187  0.000000    0.041628   
SAN LUIS                  0.018661  0.030144   0.192349  0.000000    0.022967   
SANTA CRUZ                0.058853  0.020096   0.153592  0.005742    0.050240   
SANTA FE                  0.058853  0.711979   2.072777  0.015790    0.337329   
SANTIAGO DEL ESTERO       0.014354  0.017225   0.099045  0.000000    0.012919   
TIERRA DEL FUEGO          0.038757  0.015790   0.162205  0.000000    0.031580   
TUCUMAN                   0.163640  0.101916   0.622981  0.000000    0.094739   

CAT                     PLAZA SAN MARTIN  PUERTO MADERO  RECOLETA    RETIRO  
ARGENTINA                                                                    
BUENOS AIRES                    2.091438       1.225867  0.898586  2.051245  
CATAMARCA                       0.048805       0.015790  0.024402  0.122012  
CHACO                           0.139238       0.077514  0.050240  0.330151  
CHUBUT                          0.179430       0.045934  0.045934  0.107658  
CIUDAD DE BUENOS AIRES          7.405440       3.765162  6.568578  2.160339  
CORDOBA                         2.662743       1.024905  1.174191  2.301012  
CORRIENTES                      0.248331       0.067466  0.044499  0.406230  
ENTRE RIOS                      0.304314       0.156463  0.070337  0.513888  
FORMOSA                         0.058853       0.022967  0.037321  0.147850  
JUJUY                           0.104787       0.048805  0.053111  0.140673  
LA PAMPA                        0.087562       0.017225  0.021532  0.165076  
LA RIOJA                        0.043063       0.027273  0.004306  0.093304  
MENDOZA                         0.815331       0.304314  0.357425  0.788057  
MISIONES                        0.133496       0.044499  0.027273  0.287088  
NEUQUEN                         0.406230       0.232541  0.086126  0.147850  
NO INFORMO PROVINCIA            0.001435       0.005742  0.000000  0.000000  
RIO NEGRO                       0.206704       0.061724  0.034451  0.211010  
SALTA                           0.223929       0.061724  0.081820  0.180866  
SAN JUAN                        0.124883       0.080385  0.081820  0.238283  
SAN LUIS                        0.113400       0.008613  0.030144  0.274169  
SANTA CRUZ                      0.149286       0.043063  0.045934  0.078949  
SANTA FE          